### PR TITLE
attestation extended expiry resources

### PR DESCRIPTION
**Pull request by tanay-microsoft**
## Description
</br>
 Brief description of the requested change.

## Checklist
- [ ] I have read the instructions mentioned in [Contribute to Code](/Contributing.md).
- [ ] I have read and understood the criteria described under [submitting changes](/Contributing.md#submitting-changes). 
- [ ] The title of the PR clearly describes the intent of the PR.
- [ ] This PR does not introduce any breaking changes to the code.

### CODE REVIEW

File: src/AzSK.ADO/Framework/Configurations/SVT/ControlSettings.json [hunk 1]
```diff
@@ -150,6 +150,21 @@
       "Low": 90
     }
   },
+  "ExtendedExpiryDuration": 180,
+  "ExtendedExpiryResources": [
+    {
+      "ResourceType": "ServiceConnection",
+      "ResourceIds": [
+        "organization/71a9589d-73d1-47cd-9d00-b6aee2a787ae/project/3cf66fdf-da98-4143-8ba1-7f2e141d3700/serviceconnection/e314de08-5322-4045-9093-eb24d7546783"
```
Comment (Garima-msft) [inline src/AzSK.ADO/Framework/Configurations/SVT/ControlSettings.json:158]: ExtendedExpiryResources should be empty by default

File: src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1 [hunk 2]
```diff
@@ -282,6 +282,21 @@ class SVTControlAttestation
 					$controlState.State.ApprovedExceptionID = $this.attestOptions.ApprovedExceptionID
 					$controlState.State.ExpiryDate = $exceptionApprovalExpiryDate.ToString("MM/dd/yyyy");
 				}
+
+				# Checking if the resource id is present in extended expiry list of control settings
+				if([Helpers]::CheckMember($this.ControlSettings, "ExtendedExpiryResources") -and [Helpers]::CheckMember($this.ControlSettings, "ExtendedExpiryDuration")){
+					if(($this.ControlSettings.ExtendedExpiryResources | Get-Member "ResourceType") -and ($this.ControlSettings.ExtendedExpiryResources | Get-Member "ResourceIds")) {
```
Comment (Garima-msft) [inline src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1:288]: Has this been tested in case ExtendedExpiryResources  is empty in control settings?

Comment (tanay-microsoft) [reply src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1:288]: yes, working as expected

File: src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1 [hunk 3]
```diff
@@ -405,6 +420,20 @@ class SVTControlAttestation
...
+						# Checking if the resource id is present in extended expiry list of control settings
+						if([Helpers]::CheckMember($this.ControlSettings, "ExtendedExpiryResources") -and [Helpers]::CheckMember($this.ControlSettings, "ExtendedExpiryDuration")){
+							if(($this.ControlSettings.ExtendedExpiryResources | Get-Member "ResourceType") -and ($this.ControlSettings.ExtendedExpiryResources | Get-Member "ResourceIds")) {
+								$extended_resources = $this.ControlSettings.ExtendedExpiryResources | Where { $_.ResourceType -match $controlItem.ResourceContext.ResourceTypeName }
+								# type null check
+								if($controlState.ResourceId -in $extended_resources.ResourceIds){
+									$controlState.State.ExpiryDate = $controlState.State.AttestedDate.AddDays($this.ControlSettings.ExtendedExpiryDuration);
+								}
+							}
+							else {
```
Comment (Garima-msft) [inline src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1:432]: Else not needed here. If property does not exist in control settings, it should fallback to old code

**Review by v-himkam — DISMISSED** [review]

File: src/AzSK.ADO/Framework/Configurations/SVT/ControlSettings.json [hunk 4]
```diff
@@ -150,6 +150,8 @@
       "Low": 90
     }
   },
+  "ExtendedExpiryDuration": 180,
```
Comment (SaikumarRagala) [inline src/AzSK.ADO/Framework/Configurations/SVT/ControlSettings.json:153]: ExtendedAttestationExpiryDuration

File: src/AzSK.ADO/Framework/Configurations/SVT/ControlSettings.json [hunk 5]
```diff
@@ -150,6 +150,8 @@
...
+  "ExtendedExpiryResources": [],
```
Comment (SaikumarRagala) [inline src/AzSK.ADO/Framework/Configurations/SVT/ControlSettings.json:154]: ExtendedAttestationExpiryResources

File: src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1 [hunk 6]
```diff
@@ -282,6 +282,18 @@ class SVTControlAttestation
 					$controlState.State.ApprovedExceptionID = $this.attestOptions.ApprovedExceptionID
 					$controlState.State.ExpiryDate = $exceptionApprovalExpiryDate.ToString("MM/dd/yyyy");
 				}
+
+				# Checking if the resource id is present in extended expiry list of control settings
+				if([Helpers]::CheckMember($this.ControlSettings, "ExtendedAttestationExpiryResources") -and [Helpers]::CheckMember($this.ControlSettings, "ExtendedAttestationExpiryDuration")){
+					if(($this.ControlSettings.ExtendedAttestationExpiryResources | Get-Member "ResourceType") -and ($this.ControlSettings.ExtendedAttestationExpiryResources | Get-Member "ResourceIds")) {
+						$extended_resources = $this.ControlSettings.ExtendedAttestationExpiryResources | Where { $_.ResourceType -match $controlItem.ResourceContext.ResourceTypeName }
+						# type null check
+						if($controlState.ResourceId -in $extended_resources.ResourceIds){
```
Comment (SaikumarRagala) [inline src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1:291]: This scenario will fail if you have 'ServiceConnection' in ExtendedAttestationExpiryResources, but you are scanning other resource type. please verify all scenarios.

File: src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1 [hunk 7]
```diff
@@ -405,6 +417,17 @@ class SVTControlAttestation
 							$controlState.State.ApprovedExceptionID = $this.attestOptions.ApprovedExceptionID
 							$controlState.State.ExpiryDate = $exceptionApprovalExpiryDate.ToString("MM/dd/yyyy");
 						}
+						# Checking if the resource id is present in extended expiry list of control settings
+						if([Helpers]::CheckMember($this.ControlSettings, "ExtendedAttestationExpiryResources") -and [Helpers]::CheckMember($this.ControlSettings, "ExtendedAttestationExpiryDuration")){
+							if(($this.ControlSettings.ExtendedAttestationExpiryResources | Get-Member "ResourceType") -and ($this.ControlSettings.ExtendedAttestationExpiryResources | Get-Member "ResourceIds")) {
+								$extended_resources = $this.ControlSettings.ExtendedAttestationExpiryResources | Where { $_.ResourceType -match $controlItem.ResourceContext.ResourceTypeName }
+								# type null check
+								if($controlState.ResourceId -in $extended_resources.ResourceIds){
```
Comment (SaikumarRagala) [inline src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1:425]: Same as above

File: src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1 [hunk 8]
```diff
@@ -282,6 +282,18 @@ class SVTControlAttestation
 					$controlState.State.ApprovedExceptionID = $this.attestOptions.ApprovedExceptionID
 					$controlState.State.ExpiryDate = $exceptionApprovalExpiryDate.ToString("MM/dd/yyyy");
 				}
+
+				# Checking if the resource id is present in extended expiry list of control settings
+				if([Helpers]::CheckMember($this.ControlSettings, "ExtendedAttestationExpiryResources") -and [Helpers]::CheckMember($this.ControlSettings, "ExtendedAttestationExpiryDuration")){
+					if(($this.ControlSettings.ExtendedAttestationExpiryResources | Get-Member "ResourceType") -and ($this.ControlSettings.ExtendedAttestationExpiryResources | Get-Member "ResourceIds")) {
+						$extended_resources = $this.ControlSettings.ExtendedAttestationExpiryResources | Where { $_.ResourceType -match $controlItem.ResourceContext.ResourceTypeName }
```
Comment (SaikumarRagala) [inline src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1:289]: $extended_resources -> $extendedResources

Comment (SaikumarRagala) [reply src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1:289]: Change all its occurences.

File: src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1 (hunk 7 above)
Comment (v-soukum) [reply src/AzSK.ADO/Framework/Core/SVT/SVTControlAttestation.ps1]: changes done

**Review by v-soukum — DISMISSED** [review]

**Review by SaikumarRagala — APPROVED** [review]

**Pull request merged**